src/tests/end2end/CopyTests.cpp

```diff
@@ -0,0 +1,502 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tests/NXTTest.h"
+
+#include <array>
+#include "common/Constants.h"
+#include "common/Math.h"
+#include "utils/NXTHelpers.h"
+
+class CopyTests : public NXTTest {
+    protected:
+        static constexpr unsigned int kBytesPerTexel = 4;
+
+        struct TextureSpec {
+            uint32_t width;
+            uint32_t height;
+            uint32_t x;
+            uint32_t y;
+            uint32_t copyWidth;
+            uint32_t copyHeight;
+            uint32_t level;
+        };
+
+        struct BufferSpec {
+            uint32_t size;
+            uint32_t offset;
+            uint32_t rowPitch;
+        };
+
+        BufferSpec MinimumBufferSpec(uint32_t width, uint32_t height) {
+            uint32_t rowPitch = Align(width * kBytesPerTexel, kTextureRowPitchAlignment);
+            return { rowPitch * (height - 1) + width * kBytesPerTexel, 0, rowPitch };
+        }
+
+        static void PackTextureData(const RGBA8* srcData, uint32_t width, uint32_t height, uint32_t srcTexelsPerRow, RGBA8* dstData, uint32_t dstTexelsPerRow) {
+            for (unsigned int y = 0; y < height; ++y) {
+                for (unsigned int x = 0; x < width; ++x) {
+                    unsigned int src = x + y * srcTexelsPerRow;
+                    unsigned int dst = x + y * dstTexelsPerRow;
+                    dstData[dst] = srcData[src];
+                }
+            }
+        }
+};
+
+class CopyTests_T2B : public CopyTests {
+    protected:
+        static void FillTextureData(uint32_t width, uint32_t height, uint32_t texelsPerRow, RGBA8* data) {
+            for (unsigned int y = 0; y < height; ++y) {
+                for (unsigned int x = 0; x < width; ++x) {
+                    unsigned int i = x + y * texelsPerRow;
+                    data[i] = RGBA8(x % 255, y % 255, 0, 255);
```

Comment:
Can it be RGBA8(x%256, y%256, x/256, y/256) so that we are sure each pixel is unique?